src/java/picard/analysis/replicates/MergeableMetricBase.java

```diff
@@ -0,0 +1,160 @@
+                if (equalAnnotations.length != 0) {
+                    if (!field.get(this).equals(field.get(other))) {
+                        return false;
+                    }
+                }
+            }
+        } catch (final Exception e) {
+            return false;
+        }
+        return true;
+    }
+
+    /** Merges another MergableMetricBase if possible
+     *
+     * @param other another MergableMetricBase instance to merge, must of the same class as this.
+     * @return
+     */
+    public boolean mergeIfCan(final MergeableMetricBase other) {
+        if(canMerge(other)) {
+            merge(other);
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    public void merge(final MergeableMetricBase other) {
+
+
+        for (final Field field : this.getClass().getDeclaredFields()) {
+            if(field.isSynthetic()) continue;
+
+            if (field.getAnnotationsByType(MergeByAdding.class).length +
+                    field.getAnnotationsByType(MergeByAssertEquals.class).length +
+                    field.getAnnotationsByType(NoMergingIsDerived.class).length == 0) {
+                throw new IllegalStateException("All fields of this class must be annotated with @MergeByAdding, @NoMergingIsDerived, or @MergeByAssertEquals. " +
+                        "Field " + field.getName() + " isn't annotated.");
+            }
+
+            final Annotation[] summableAnnotations = field.getAnnotationsByType(MergeByAdding.class);
+            if (summableAnnotations.length != 0) {
+                try {
+
+                    if (field.getType() == Integer.class) {
+                        field.set(this, (Integer) field.get(this) + (Integer) field.get(other));
+                    } else if (field.getType() == int.class) {
+                        field.set(this, (int) field.get(this) + (int) field.get(other));
+                    } else if (field.getType() == Float.class) {
+                        field.set(this, (Float) field.get(this) + (Float) field.get(other));
+                    } else if (field.getType() == float.class) {
+                        field.set(this, (float) field.get(this) + (float) field.get(other));
+                    } else if (field.getType() == Double.class) {
+                        field.set(this, (Double) field.get(this) + (Double) field.get(other));
+                    } else if (field.getType() == double.class) {
+                        field.set(this, (double) field.get(this) + (double) field.get(other));
+                    } else if (field.getType() == Long.class) {
+                        field.set(this, (Long) field.get(this) + (Long) field.get(other));
+                    } else if (field.getType() == long.class) {
+                        field.set(this, (long) field.get(this) + (long) field.get(other));
+                    } else if (field.getType() == Byte.class) {
```

Comment:
Yeah, I discussed this with Louis and David and they didn't have any better
solution....

On Sat, Apr 16, 2016 at 10:09 PM, Eric Banks <notifications@github.com>
wrote:

> In src/java/picard/analysis/replicates/MergeableMetricBase.java
> <https://github.com/broadinstitute/picard/pull/482#discussion_r59977241>:
>
> > +                        field.set(this, (Integer) field.get(this) + (Integer) field.get(other));
> > +                    } else if (field.getType() == int.class) {
> > +                        field.set(this, (int) field.get(this) + (int) field.get(other));
> > +                    } else if (field.getType() == Float.class) {
> > +                        field.set(this, (Float) field.get(this) + (Float) field.get(other));
> > +                    } else if (field.getType() == float.class) {
> > +                        field.set(this, (float) field.get(this) + (float) field.get(other));
> > +                    } else if (field.getType() == Double.class) {
> > +                        field.set(this, (Double) field.get(this) + (Double) field.get(other));
> > +                    } else if (field.getType() == double.class) {
> > +                        field.set(this, (double) field.get(this) + (double) field.get(other));
> > +                    } else if (field.getType() == Long.class) {
> > +                        field.set(this, (Long) field.get(this) + (Long) field.get(other));
> > +                    } else if (field.getType() == long.class) {
> > +                        field.set(this, (long) field.get(this) + (long) field.get(other));
> > +                    } else if (field.getType() == Byte.class) {
>
> This is all kind of ugly. I'd like to ask others whether there's a cleaner
> way of doing this. Okay?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/482/files/dc932b88e1f17c278b4e71abff7593cbc904e5a3#r59977241>
>